src/prompt/variable.rs

```diff
@@ -16,31 +35,200 @@ use rustyline::hint::Hinter;
+        ParsingError::Mistake { kind: Some(kind), description: message }
+    }
+
+    // if combining multiple errors, we show them one after the other
+    fn append(input: &str, kind: ErrorKind, other: Self) -> Self {
+        let message = format!("{}, then: {}", other, ParsingError::from_error_kind(input, kind));
+        ParsingError::Mistake { kind: Some(kind), description: message }
+    }
+
+    fn from_char(input: &str, c: char) -> Self {
+
+        let message = if input != "" {
+            format!("Expected '{}' in {:?}", c, input)
+        } else {
+            format!("Expected '{}'", c)
+        };
+        ParsingError::Mistake { kind: None, description: message }
+    }
+
+    fn or(self, other: Self) -> Self {
+        let message = format!("{}, or: {}", self, other);
+        ParsingError::Mistake { kind: None, description: message }
+    }
+}
+
+impl ContextError<&str> for ParsingError {
+    fn add_context(_input: &str, ctx: &'static str, other: Self) -> Self {
+        let message = format!("{} -> {}", ctx, other);
+        ParsingError::Mistake { kind: None, description: message }
+    }
+}
+
+impl FromExternalError<&str, anyhow::Error> for ParsingError {
+    fn from_external_error(input: &str, kind: ErrorKind, e: anyhow::Error) -> Self {
+        ParsingError::External {
+            error: e,
+            kind,
+            description: format!("Failed at '{}'", input)
+        }
+    }
+}
+
+#[repr(u8)]
+pub enum InputFlags {
+    None,
+    ForceQuotedStrings
 }
 
 pub trait VariableInput: fmt::Debug + Send + Sync + 'static {
-    fn parse(&self, input: &str) -> Result<Value, Error>;
     fn type_name(&self) -> &str;
+    fn parse<'a >(&self, input: &'a str, flags: InputFlags) -> ParseResult<'a>;
+}
+
+fn white_space<'a, O, E: ParseError<&'a str>, F: Parser<&'a str, O, E>>(
+    f: F,
+) -> impl Parser<&'a str, O, E> {
+    delimited(multispace0, f, multispace0)
+}
+fn space(i: &str) -> IResult<&str, &str, ParsingError> {
+    let chars = " \t\r\n";
+    take_while(move |c| chars.contains(c))(i)
+}
+
+fn quoted_str(input: &str) -> IResult<&str, String, ParsingError> {
+    context(
+        "any_quote_str",
+        alt((
+            single_quoted_str,
+            double_quoted_str
+        ))
+    )(input)
+}
+
+fn single_quoted_str(input: &str) -> IResult<&str, String, ParsingError> {
+    context(
+        "single_quote_str",
+        |s| quoted_str_parser(s, '\'', "\'\\")
+    )(input)
+}
+
+fn double_quoted_str(input: &str) -> IResult<&str, String, ParsingError> {
+    context(
+        "double_quote_str",
+        |s| quoted_str_parser(s, '\"', "\"\\")
+    )(input)
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum StringFragment<'a> {
+    Literal(&'a str),
+    EscapedChar(char),
+    EscapedWS,
+}
+
+// heavily based on https://github.com/rust-bakery/nom/blob/main/examples/string.rs
+fn quoted_str_parser<'a>(input: &'a str, quote: char, esc: &str) -> IResult<&'a str, String, ParsingError> {
```

Comment:
I think the easiest (and best) way to do this would be to take advantage of some existing code in our tokenizer for this: https://github.com/edgedb/edgedb/blob/62d8fcfcbeded747106a8914e1b4ef4bac37aa61/edb/edgeql-parser/src/helpers/strings.rs#L63-L168

So I think the approach then would be to match a quoted string *uninterpreted* (only understanding escapes as much as is needed to allow an escaped quote), and then passing that uninterpreted quoted string to unquote_string)

Reply:
(That is what we do in our actual lexer)

Reply:
<img width="378" alt="image" src="https://github.com/edgedb/edgedb-cli/assets/49576606/e6d071fd-40e8-4543-a358-427a7156ab30">
I've updated the parser to follow the same rules of unicodes and such, I tried integrating `unquote_string` but I failed at it. Maybe if you already know how to integrate it you can.

Reply:
Could you elaborate on what the snags were? What I would probably recommend is writing a combinator that matches a quoted string but doesn't really worry about what the return value is or if the escapes are valid; then you can use that with `recognize`, and pass the result of that to `unquote_string`

Reply:
https://github.com/edgedb/edgedb/blob/master/edb/edgeql-parser/src/tokenizer.rs#L659-L669 is the tokenizer code for matching a quoted string; one option (though not the most idiomatic for parser combinators) is to just write a combinator directly that does that logic

Reply:
> Could you elaborate on what the snags were?

Getting it to match a quoted string while ignoring other escape sequences, ex `\x33` or `\u236A`, kept either overrunning the quoted string (greedily) or not matching anything at all because of non recognizable escapes

Reply:
What did the combinators you were producing look like?
I *think* that for single quote strings, it should work to match a sequence of:
 * any non `'` non `\` character
 * `\` followed by any other character

Reply:
something like:
```rs
    context(
        "quoted_string",
        delimited(
            char(quote),
            map_res(
                fold_many0(
                    alt((
                        map(
                            verify(is_not(esc), |s: &str| !s.is_empty()),
                            StringFragment::Literal
                        ),
                        map(
                            preceded(
                                char('\\'),
                                alt((
                                    value('\\', char('\\')),
                                    value('/', char('/')),
                                    value('"', char('"')),
                                    value('\'', char('\'')),
                                ))
                            ),
                            StringFragment::EscapedChar
                        ),
                        value(
                            StringFragment::EscapedWS,
                            preceded(char('\\'), nom::character::streaming::multispace1)
                        ),
                    )),
                    String::new,
                    |mut string, fragment| {
                        match fragment {
                            StringFragment::Literal(s) => string.push_str(s),
                            StringFragment::EscapedChar(c) => string.push(c),
                            StringFragment::EscapedWS => {}
                        }
                        string
                    }
                ),
                |s| unquote_string(&s).map(|v| v.into()).context("Failed to unquote string")
            )
            char(quote),
        )
    )(input)
```

Reply:
What is `esc` here? The relevant quote and backslash?

One problem here is that `unquote_string` expects to get the full quoted string, *including* the quote characters. I also think the escape handling is busted (it doesn't match all escape characters (I think) and it *interprets* some of the escape characters, which will do the wrong thing once passed to unquote_string).

I think the best approach remains the one I outlined above: write a combinator that *recognizes* a quoted string without interpreting it at all (it should probably just produce `()` as its value), then use `recognize` to get the string out and call `unquote_string`.

Reply:
You can follow up with me on slack if you have any questions

Reply:
I think I got a working impl using `unquote_string`